### DISCUSSION TITLE
Will the any-maze TTL cable work to send a TTL signal from Bonsai?

### DISCUSSION THREAD

**Discussion by MrNerdly** [bonsai-rx/bonsai#1713] [Q&A] [answered]
Hello,
I am trying to align a video recording with fiber photometry data, and one way I've found to do so is by sending a TTL pulse to the fiber photometry system when the video begins and ends recording. ANY-maze can apparently do this built in using a cable ( https://stoeltingco.com/Neuroscience/TTL-Cables~10288 ), but I would like to use Bonsai. 

Would Bonsai be able to send the signal across this cable? Or do I need to purchase a data acquisition system and BNC cable? Thank you in advance for any insight!

**Comment by bruno-f-cruz** [comment]
No idea what this cable is, but depending what you mean by "alignment" you can probably use an arduino to send the TTL pulse and save some money. That being said, sending a single pulse from your software to get stamped by your hardware (i.e. fiber photometry rig) will not provide precise video alignment as there might be substantial drift / jitter between the two systems. You should look into other solutions like sending the same TTL once in a while to both the camera and the fiber photometry system.

**Comment by MrNerdly** [comment]
Thanks for the input! Can the arduino send that input to a usb webcam? I originally considered an arduino setup, but noticed papers using this cable but within the proprietary any-maze ecosystem.

**Reply by bruno-f-cruz** [reply]
The Arduino can send a 5V pulse in the same way I understand aforementioned cable can. If you can add a circuit diagram of what you are trying to achieve I might be able to provide more help, but at the end of the day a 5V pulse is still a 5V pulse, it shouldn't matter wether its delivered via a proprietary cable (unfortunately, the details of what this cable actually does are scarce at best) or an arduino...

**Reply by MrNerdly** [reply]
![circuit_diagram_fiberpho](https://github.com/bonsai-rx/bonsai/assets/162047594/f236ca60-23e1-404d-9a14-9bf3bbd4c961)
I've attached a rather crude circuit diagram, but in essence we just want to collect video data in time with the fiber photometry data so we can later look at 3 second sections of the fiber photometry data based on the video data. What we have so far is the fiber pho system that as DNC in and out and a USB camera. We were told to buy a Data acquisition system to a TTL pulse that turns on the fiber photometry system in time with the camera. However I read of a paper using an arduino instead and another that had a start and stop pulse for when the video recording started and stopped. This TTL pulse was sent to the fiber photometry system.

So, would we be able to achieve this goal with an arduino, the fiber photometry system with a BNC connection, and a usb camera? If we need your initial suggestion of periodic TTL pulses to both devices, can those even be sent to a usb camera? Thank you!

**Reply by bruno-f-cruz** [reply]
I am not sure I fully understand what you mean by synchronization after looking at your last post. If you have a USB webcam I assume you won't be able to trigger frames or inject an external TTL into it, so what do you want the TTL for anyway? On top of this, what is the level of synchronization you are looking for? sub millisecond? sub 100ms? just knowing when the camera started in relation to the fiber photometry and you don't care a about a potential drift? Depending on the answer, the solution may be different!

**Reply by MrNerdly** [reply]
I see what you mean about how the answer can vary!

Synchrony was probably a poor word choice. We are aiming for alignment between what is happening in a given video frame and the fiber photometry data. Looking at these other software's that use Data acquisition system hardware, it looks like a TTL pulse from when the video recording started and another when the video recording ends. The fiber photometry equipment can receive these pulses and place them in a second channel showing where they occurred in relation to the fiber photometry data. This let them have 2 points for alignment. We end up looking at about 5 seconds of fiber photometry data with the middle point of that being based on what we saw of interest in the video data, so I imagine we're in the sub 100 ms, maybe sub 50ms? 

 As far as I know our USB camera does not on it's own have the ability to interact with or send TTL pulses. 
 Does that add enough clarity? My background is in behavior and chemogenetic manipulations, so while I know what I want to get from this setup at the end, I'm not quite used to describing or thinking about it in this perspective.

**Reply by bruno-f-cruz** [reply, marked as answer]
I am still a bit confused. I might be missing something about the rig/software you are using. How does the software know when the trigger was sent relative to the camera starting? They are independent systems, you will once again have large amounts of jitter and I doubt you will reach the 50ms benchmark you mentioned. If you want to use a usb camera, the best solution I can think of is to flash an led every once in a while on the camera and send the same pulse to your photometry system. At least this way you will have several hardware-level synchronization events you can use.

**Reply by MrNerdly** [reply]
It does sound like, from multiple sources, an LED connected to a circuit that also sends a TTL pulse is the best answer, thank you for your help!

**Comment by jfrazao** [comment]
The only way to have a cheap USB camera aligned is to have a LED in the
field of view of the camera, this LED is wired to whichever TTL pulse you
use for sinchronization:  the start and end of the trial; Or Random pulses;
or start and end of stimulation; etc.
Sometimes people use infrared led, so the animal cannot see them. USB
cameras usually are cheap enough that their bad IR cut filters are
precisely fit for this.

**Reply by MrNerdly** [reply]
Hello, Thanks for this information! This sounds like the best option.
